gno/p/role_manager/role_manager.gno

```diff
@@ -0,0 +1,187 @@
+package role_manager
+
+import (
+	"std"
+
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/ownable"
+)
+
+const (
+	RoleUser  = "user"
+	RoleAdmin = "admin"
+)
+
+const (
+	PermissionAll       = "all"
+	PermissionWriteRole = "write:role"
+)
+
+type Role struct {
+	name        string
+	permissions *avl.Tree // permission -> struct{}
+	users       *avl.Tree // user addr -> struct{}
+}
+
+type RoleManager struct {
+	owner *ownable.Ownable
+	roles *avl.Tree // role name -> Role
+	users *avl.Tree // user addr -> role[]
+}
+
+func NewWithAddress(addr std.Address) *RoleManager {
+	rm := &RoleManager{
+		owner: ownable.NewWithAddress(addr),
+		roles: avl.NewTree(),
+		users: avl.NewTree(),
+	}
+
+	userRole := &Role{
+		name:        RoleUser,
+		permissions: avl.NewTree(),
+		users:       avl.NewTree(),
+	}
+
+	adminRole := &Role{
+		name:        RoleAdmin,
+		permissions: avl.NewTree(),
+		users:       avl.NewTree(),
+	}
+	adminRole.permissions.Set(PermissionAll, struct{}{})
+
+	rm.roles.Set(RoleUser, userRole)
+	rm.roles.Set(RoleAdmin, adminRole)
+
+	return rm
+}
```

Comment:
I'm not a fan of forcing the creation of these roles

Can we have a way to instantiate a blank RoleManager with just the owner/admin?

Reply:
let's talk about this, i can understand for admin role but i like the fact that everyone is a basic user so if you call a function hasRole on a new addr it will act as an user and not as "this user does not exist". In this way you can affect some rights or roles to user that never interact before with your realms or you can let the users role without any permissions if you don't want to do this.

Reply:
> In this way you can affect some rights or roles to user that never interact before with your realms
I think this is a bad pattern, it breaks least permissions principle

but I understand some use case, if there is something you only allow to members then want at some point allow to everyone, mmh..
I'm not sure we should support it. another principle of security system is to prevent bad usage (shooting itself in the foot)